### PR TITLE
Added two examples - One to configure NAT instance for Internet access from private instances, another for a DNS VM to allow on-premises network to query DNS hostnames in the VCN

### PR DESCRIPTION
**Pull request by sivaedupuganti-oracle**
Added example to configure NAT instance for Internet access from private instances

### CODE REVIEW

File: docs/examples/networking/nat/vcn.tf [hunk 1]
```diff
@@ -0,0 +1,215 @@
...
+}
+variable "InstanceOSVersion" {
+    default = "7.4"
+}
+provider "oci" {
+    tenancy_ocid = "${var.tenancy_ocid}"
+    user_ocid = "${var.user_ocid}"
+    fingerprint = "${var.fingerprint}"
+    private_key_path = "${var.private_key_path}"
+    private_key_password = "${var.private_key_password}"
```
Comment (briangustafson) [inline docs/examples/networking/nat/vcn.tf]: Need to define this as a variable.

File: docs/examples/networking/nat/vcn.tf [hunk 2]
```diff
@@ -0,0 +1,215 @@
+variable "tenancy_ocid" {}
```
Comment (briangustafson) [inline docs/examples/networking/nat/vcn.tf]: This file should probably be called nat.tf - seems to be the pattern we're using.

File: docs/examples/networking/nat/user_data.tpl [hunk 3]
```diff
@@ -0,0 +1,11 @@
+write_files:
+  - path: /etc/sysctl.d/98-ip-forward.conf
+    content: |
+      net.ipv4.ip_forward = 1
+
+runcmd:
```
Comment (briangustafson) [inline docs/examples/networking/nat/user_data.tpl]: Would be useful to add a comment on what this is doing (for those not so familiar with these tools, such as myself).

File: docs/examples/networking/nat/env-vars [hunk 4]
```diff
@@ -0,0 +1,8 @@
+### Region
```
Comment (briangustafson) [inline docs/examples/networking/nat/env-vars]: This file should have all of the variables used.

File: docs/examples/networking/nat/env-vars [hunk 5]
```diff
@@ -0,0 +1,8 @@
...
+export TF_VAR_region=""
+
+### Compartment
+export TF_VAR_compartment_ocid=""
+
+export TF_VAR_ssh_public_key=$(cat .../id_rsa.pub)
```
Comment (briangustafson) [inline docs/examples/networking/nat/env-vars]: How about the follow, copied from another example:

export TF_VAR_ssh_public_key=$(cat <path to public key>)
export TF_VAR_ssh_private_key=$(cat <path to private key>)

File: docs/examples/networking/nat/vcn.tf [hunk 6]
```diff
@@ -0,0 +1,215 @@
...
+        icmp_options {
+            "type" = 3
+            "code" = 4
+        }
+    }]
+}
+
+resource "oci_core_subnet" "MgmtSubnet" {
+    availability_domain = "${lookup(data.oci_identity_availability_domains.ADs.availability_domains[var.AD - 1],"name")}"
+    cidr_block = "10.0.0.0/24"
```
Comment (briangustafson) [inline docs/examples/networking/nat/vcn.tf]: The VCN CIDR has a variable, but the Subnet CIDRs do not. I would think that if one is a variable then the others should be as well.

File: docs/examples/networking/nat/vcn.tf [hunk 7]
```diff
@@ -0,0 +1,215 @@
...
+    user_ocid = "${var.user_ocid}"
+    fingerprint = "${var.fingerprint}"
+    private_key_path = "${var.private_key_path}"
+    private_key_password = "${var.private_key_password}"
+    region = "${var.region}"
+}
+data "oci_identity_availability_domains" "ADs" {
+    compartment_id = "${var.tenancy_ocid}"
+}
+variable "VCN-CIDR" {
```
Comment (briangustafson) [inline docs/examples/networking/nat/vcn.tf]: Move to top with other vars.

File: docs/examples/networking/nat/nat.tf [hunk 8]
```diff
@@ -0,0 +1,226 @@
...
+variable "vcn_cidr" {
+    default = "10.0.0.0/16"
+}
+
+variable "mgmt_subnet_cidr" {
+    default = "10.0.0.0/16"
+}
+
+variable "private_subnet_cidr" {
+    default = "10.0.0.0/16"
```
Comment (briangustafson) [inline docs/examples/networking/nat/nat.tf]: When running using these defaults I'm getting the error below - I think this needs to be a different value from the other subnet:

* oci_core_subnet.PrivateSubnet: Status: 400; Code: InvalidParameter; OPC Request ID: /87AF334F38C76347EE6B866A2F26EA55/943BBE5065234BEE81B795DD3DFC2518; Message: The requested CIDR 10.0.0.0/16 is invalid: subnet ocid1.subnet.oc1.iad.aaaaaaaa4fdarrsa7jyqy7wrs5athfpct55dea2n5cntqypu4grd5fehjjwa with CIDR 10.0.0.0/16 overlaps with this CIDR.

**Comment by sivaedupuganti-oracle** [comment]
@briangustafson fixed the errors with variables. Updated the hybrid DNS example to launch two DNS VMs in separate ADs, and added a script to modify default DHCP options.

File: docs/examples/networking/nat/user_data.tpl [hunk 9]
```diff
@@ -0,0 +1,18 @@
+#
```
Comment (briangustafson) [inline docs/examples/networking/nat/user_data.tpl]: What's the simplest way to verify that this is working correctly?

Comment (sivaedupuganti-oracle) [reply docs/examples/networking/nat/user_data.tpl]: Addressed this in the README.

**Comment by sivaedupuganti-oracle** [comment]
Updated hybrid_dns to remove the steps to update default DHCP options and left a note in Readme instead. Moved docx to markdown.

File: docs/examples/networking/hybrid_dns/Hybrid-DNS-configuration-using-DNS-VM-in-VCN.md [hunk 10]
```diff
@@ -0,0 +1,78 @@
...
+
+
+2. Create a subnet, which uses the security list and DHCP options set created above.
+3. Launch a VM with latest &#39;Oracle Linux 7.4&#39; image into this subnet
+4. Install &amp; Configure named
+   * $ sudo yum install bind
+   * $ sudo firewall-cmd --permanent --add-port=53/udp
+   * $ sudo firewall-cmd --permanent --add-port=53/tcp
+   * $ sudo /bin/systemctl restart firewalld
+   * $ cat &gt; /etc/named.conf
```
Comment (briangustafson) [inline docs/examples/networking/hybrid_dns/Hybrid-DNS-configuration-using-DNS-VM-in-VCN.md]: The formatting for this does not show up correctly. (The the "view" button for this file in the upper right.)

File: docs/examples/networking/hybrid_dns/Hybrid-DNS-configuration-using-DNS-VM-in-VCN.md [hunk 11]
```diff
@@ -0,0 +1,78 @@
+# Hybrid DNS Configuration
+
+Oracle Cloud Infrastructure (OCI) customers can configure DNS names for their instances in the Virtual Cloud Network (VCN) as described in [DNS in Your Virtual Cloud Network](https://docs.us-phoenix-1.oraclecloud.com/Content/Network/Concepts/dns.htm). The DNS names are resolvable only within the VCN using the VCN DNS resolver available at 169.254.169.254 IP address. This IP address is only reachable from instances in the VCN.
```
Comment (briangustafson) [inline docs/examples/networking/hybrid_dns/Hybrid-DNS-configuration-using-DNS-VM-in-VCN.md]: "available at IP address 169.254.169.254"

File: docs/examples/networking/nat/README.md [hunk 12]
```diff
@@ -0,0 +1,29 @@
...
+The public instance is configured as a NAT instance (by enabling forwarding and configuring firewall to do forwarding/masquerading).
+The private subnet's route table is configured to use the NAT instance's private IP address as the default route target. See [Using a Private IP as a Route Target](https://docs.us-phoenix-1.oraclecloud.com/Content/Network/Tasks/managingroutetables.htm#privateip) for more details on this feature.
+
+### Using this example
+* Update env-vars with the required information. Most examples use the same set of environment variables so you only need to do this once.
+* Source env-vars
+  * `$ . env-vars`
+* Update variables in vcn.tf as applicable to your target environment.
+
+Once the environment is built, the private instance has Internet connectivity even when it doesn't have a public IP address and it's subnet's route table doesn't contain Internet gateway. You can login into the private instance (from the nat instance) and then run a command like 'ping oracle.com' to verify connectivity.
```
Comment (briangustafson) [inline docs/examples/networking/nat/README.md]: "and its subnet's route table doesn't contain an Internet gateway."

**Comment by sivaedupuganti-oracle** [comment]
@briangustafson this PR contains two major scenarios that can be implemented in networking. I'm interested to get this in front of people who want to use this. Let's push this version. Any further feedback can be iterated in a future version, at which time there will be feedback at broader level from other sources too.

**Pull request merged**